src/python/pants/backend/python/packaging/pyoxidizer/rules.py

```diff
@@ -48,9 +53,36 @@ class PyOxidizerFieldSet(PackageFieldSet):
     template: PyOxidizerConfigSourceField
 
 
+@dataclass(frozen=True)
+class PyoxidizerRunnerScript:
+    digest: Digest
+    path: str
+
+
+@rule
+async def create_pyoxidizer_runner_script() -> PyoxidizerRunnerScript:
+    # Note: PyOxidizer expects an absolute path for its cache dir, which can only be resolved
+    # from within the execution sandbox. Thus, this code uses a bash script to be able to resolve
+    # absolute paths inside the sandbox.
+    script = FileContent(
+        "__run_pyoxidizer.sh",
+        dedent(
+            f"""\
+            export PYOXIDIZER_CACHE_DIR="$(/bin/pwd)/.cache/pyoxidizer"
```

Comment:
No trepidation here? I had some and found means to only rely on bash to find absolute oaths in the venv wrapper scripts.

Reply:
Copy pasta from Go:

https://github.com/pantsbuild/pants/blob/75c6ab0435b584120abbe82d06778ea9f97e679a/src/python/pants/backend/go/util_rules/sdk.py#L66-L88

That bio-inspired computing class I told you about 3 years ago mentioned a theory about how programming is made up of "memes" that get shared like this, little is original. Iirc, similar to how biological and cultural evolution both work.

> I found means to only rely on bash to find absolute oaths in the venv wrapper scripts.

Ah. I will use that and try to generalize it, e.g. update Go to use it.

Reply:
I tried implementing this but am seeing `BASH_SOURCE` set to `.` instead of the absolute path. The part about the engine absolutifying argv0 doesn't seem to be working for me. Still investigating, but decided to do this in a followup where I can write a test.

```diff
diff --git a/src/python/pants/backend/python/packaging/pyoxidizer/rules.py b/src/python/pants/backend/python/packaging/pyoxidizer/rules.py
index 9cfa0ad06..cff6c9ade 100644
--- a/src/python/pants/backend/python/packaging/pyoxidizer/rules.py
+++ b/src/python/pants/backend/python/packaging/pyoxidizer/rules.py
@@ -8,6 +8,7 @@ import logging
 import os
 from dataclasses import dataclass
 from textwrap import dedent
+from typing import ClassVar
 
 from pants.backend.python.packaging.pyoxidizer.config import PyOxidizerConfig
 from pants.backend.python.packaging.pyoxidizer.subsystem import PyOxidizer
@@ -27,7 +28,7 @@ from pants.engine.fs import (
     MergeDigests,
     Snapshot,
 )
-from pants.engine.process import BashBinary, Process, ProcessResult
+from pants.engine.process import ENSURE_ABSOLUTE_BASH_FUNCTION, BashBinary, Process, ProcessResult
 from pants.engine.rules import Get, MultiGet, collect_rules, rule
 from pants.engine.target import (
     DependenciesRequest,
@@ -58,11 +59,11 @@ class PyoxidizerRunnerScript:
     digest: Digest
     path: str
 
-    CACHE_PATH = os.path.join(".cache", "pyoxidizer")
+    CACHE_PATH: ClassVar[str] = os.path.join(".cache", "pyoxidizer")
 
 
 @rule
-async def create_pyoxidizer_runner_script() -> PyoxidizerRunnerScript:
+async def create_pyoxidizer_runner_script(bash: BashBinary) -> PyoxidizerRunnerScript:
     # Note: PyOxidizer expects an absolute path for its cache dir, which can only be resolved
     # from within the execution sandbox. Thus, this code uses a bash script to be able to resolve
     # absolute paths inside the sandbox.
@@ -70,10 +71,17 @@ async def create_pyoxidizer_runner_script() -> PyoxidizerRunnerScript:
         "__run_pyoxidizer.sh",
         dedent(
             f"""\
-            export PYOXIDIZER_CACHE_DIR="$(/bin/pwd)/{PyoxidizerRunnerScript.CACHE_PATH}"
+            #!{bash.path}
+
+            {ENSURE_ABSOLUTE_BASH_FUNCTION}
+
+            export PYOXIDIZER_CACHE_DIR="$(ensure_absolute {PyoxidizerRunnerScript.CACHE_PATH})"
             exec "$@"
             """
         ).encode("utf-8"),
+        is_executable=True,
     )
     digest = await Get(Digest, CreateDigest([script]))
     return PyoxidizerRunnerScript(digest, script.path)
@@ -84,7 +92,6 @@ async def package_pyoxidizer_binary(
     pyoxidizer: PyOxidizer,
     field_set: PyOxidizerFieldSet,
     runner_script: PyoxidizerRunnerScript,
-    bash: BashBinary,
 ) -> BuiltPackage:
     direct_deps, pyoxidizer_pex = await MultiGet(
         Get(Targets, DependenciesRequest(field_set.dependencies)),
@@ -163,7 +170,7 @@ async def package_pyoxidizer_binary(
     )
     process_with_caching = dataclasses.replace(
         pex_process,
-        argv=(bash.path, runner_script.path, *pex_process.argv),
+        argv=(f"./{runner_script.path}", *pex_process.argv),
         append_only_caches={
             **pex_process.append_only_caches,
             "pyoxidizer": runner_script.CACHE_PATH,
diff --git a/src/python/pants/backend/python/util_rules/pex.py b/src/python/pants/backend/python/util_rules/pex.py
index 632adf61a..c793323cd 100644
--- a/src/python/pants/backend/python/util_rules/pex.py
+++ b/src/python/pants/backend/python/util_rules/pex.py
@@ -58,7 +58,13 @@ from pants.engine.fs import (
 from pants.engine.internals.native_engine import Snapshot
 from pants.engine.internals.selectors import MultiGet
 from pants.engine.platform import Platform
-from pants.engine.process import BashBinary, Process, ProcessCacheScope, ProcessResult
+from pants.engine.process import (
+    ENSURE_ABSOLUTE_BASH_FUNCTION,
+    BashBinary,
+    Process,
+    ProcessCacheScope,
+    ProcessResult,
+)
 from pants.engine.rules import Get, collect_rules, rule
 from pants.engine.target import HydratedSources, HydrateSourcesRequest, SourcesField, Targets
 from pants.util.frozendict import FrozenDict
@@ -624,23 +630,7 @@ class VenvScriptWriter:
             #!{bash.path}
             set -euo pipefail
 
-            # N.B.: We convert all sandbox root relative paths to absolute paths so this script
-            # works when run with a cwd set elsewhere.
-
-            # N.B.: This relies on BASH_SOURCE which has been available since bash-3.0, released in
-            # 2004. In turn, our use of BASH_SOURCE relies on the fact that this script is executed
-            # by the engine via its absolute path.
-            ABS_SANDBOX_ROOT="${{BASH_SOURCE%/*}}"
-
-            function ensure_absolute() {{
-                local value0="$1"
-                shift
-                if [ "${{value0:0:1}}" == "/" ]; then
-                    echo "${{value0}}" "$@"
-                else
-                    echo "${{ABS_SANDBOX_ROOT}}/${{value0}}" "$@"
-                fi
-            }}
+            {ENSURE_ABSOLUTE_BASH_FUNCTION}
 
             export {" ".join(env_vars)}
             export PEX_ROOT="$(ensure_absolute ${{PEX_ROOT}})"
diff --git a/src/python/pants/engine/process.py b/src/python/pants/engine/process.py
index 35a24316c..25145e731 100644
--- a/src/python/pants/engine/process.py
+++ b/src/python/pants/engine/process.py
@@ -676,5 +676,28 @@ async def find_binary(request: BinaryPathRequest) -> BinaryPaths:
     )
 
 
+"""Defines a bash function `ensure_absolute` to absolutify the input path.
+
+N.B.: This relies on BASH_SOURCE which has been available since bash-3.0, released in
+2004. In turn, our use of BASH_SOURCE relies on the fact that this script is executed
+by the engine via its absolute path.
+"""
+ENSURE_ABSOLUTE_BASH_FUNCTION = dedent(
+    """\
+    ABS_SANDBOX_ROOT="${BASH_SOURCE%/*}"
+
+    function ensure_absolute() {
+        local value0="$1"
+        shift
+        if [ "${value0:0:1}" == "/" ]; then
+            echo "${value0}" "$@"
+        else
+            echo "${ABS_SANDBOX_ROOT}/${value0}" "$@"
+        fi
+    }
+    """
+)
+
+
 def rules():
     return collect_rules()
```

Reply:
That's just one way to not cheat. Another is to request the binary path.